### PR TITLE
changed w_t to buffer

### PR DESCRIPTION
**Pull request by Olivierj23**
w_t is now a buffer rather than parameter

## Description
Addressed changing task weights to a buffer instead of a parameter

## Relevant issues
#657

### CODE REVIEW

File: chemprop/models/model.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def __init__(
             else [self.predictor._default_metric, self.criterion]
         )
         w_t = torch.ones(self.n_tasks) if w_t is None else torch.tensor(w_t)
-        self.w_t = nn.Parameter(w_t.unsqueeze(0), False)
+        self.w_t = self.register_buffer("w_t", w_t.unsqueeze(0), persistent=False)
```
Comment (davidegraff) [inline chemprop/models/model.py:94]:
```suggestion
        self.register_buffer("w_t", w_t.unsqueeze(0), persistent=False)
```
`register_buffer()` modifies the object in-place. Also, is there a reason `persistent=False`? Not storing this in the state dict means that a model checkpoint doesn't save how it was trained.

Comment (Olivierj23) [reply chemprop/models/model.py]: Sorry, I am new to working on chemprop. I did not read the documentation closely enough for the function. I have committed the fix now.

Comment (am2145) [reply chemprop/models/model.py]: I think the `persistent=False` is because `w_t` is not required for inference, so it isn't kept in the .pt file's state dict. The input task weights are saved in the .ckpt file as hyperparameters so we should still be able to re-initialize/re-create how the model is trained.

Comment (davidegraff) [reply chemprop/models/model.py]: no worries- thanks for the PR!

Comment (KnathanM) [reply chemprop/models/model.py]: The task weights are saved in the .ckpt file when we call `self.save_hyperparameters()`. This means that if you don't specify it, it is saved as `"w_t": None`. You'd have to know that this means all tasks are equally weighted.

Comment (davidegraff) [reply chemprop/models/model.py]: perhaps then we can add `"w_t"` to the "ignore" list and make this buffer persistent?

Comment (KnathanM) [reply chemprop/models/model.py]: If we want `w_t` in the `state_dict`, I can add it to ignore and make it persistent. But in my understanding, the `state_dict` is for values that are needed for inference, which `w_t` is not. The [docs](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer) use BatchNorm's `running_mean` as an example of a buffer that should be persistent. The running mean is used during inference. I think `w_t` is more similar to the learning rate, a hyperparameter that affects only training which is saved using `self.save_hyperparameters()` so that it is available in a `.ckpt` file to restart training.

Comment (KnathanM) [reply chemprop/models/model.py]: Should the learning rates also go in the `state_dict`? If not, what is the difference between `w_t` and e.g. `init_lr`? Both control how big of steps to take in the direction of the gradients so are only used during training. It makes sense to me that both are saved in model.hparams instead of the `state_dict`. The `state_dict` belongs to torch (which is training unaware) and lightning doesn't touch it, other than inheriting the related methods from `Module` in `LightningModule`.

Comment (davidegraff) [reply chemprop/models/model.py]: if you can tell me with good reason that the learning rates are valid buffers from the model's perspective, then by all means place them in there. The state dict belongs to torch, but its purpose is solely as a serialization container for parameters and buffers, the meaning of which _we_ define based on our use-case.

But bigger than this is the question of why `w_t` is even inside the MPNN in the first case. It makes more sense to me for it to be a buffer inside the `LossFunction` or `Predictor`.

Comment (KnathanM) [reply chemprop/models/model.py]: I agree and think it makes the most sense to put `w_t` inside LossFunction because a MSELoss that has `w_t = [1,2]` computes a different loss value than the standard MSELoss. The rest of the arguments in `LossFunction.__call__()` are potentially different for each datapoint. I have made these changes in #779. In this approach, `w_t` shows up in the model files under `d["hyper_parameters"]["predictor"]["criterion"].w_t`.

**Comment by KnathanM** [comment]
Superseded by #779
